src/wallet/wallet2.cpp

```diff
@@ -12901,6 +12902,74 @@ void wallet2::throw_on_rpc_response_error(const boost::optional<std::string> &st
   THROW_WALLET_EXCEPTION_IF(*status != CORE_RPC_STATUS_OK, tools::error::wallet_generic_rpc_error, method, m_trusted_daemon ? *status : "daemon error");
 }
 //----------------------------------------------------------------------------------------------------
+
+bool wallet2::save_to_file(const std::string& path_to_file, const std::string& raw) const
+{
+  if (m_export_format == ExportFormat::Hex)
+  {
+    std::string h = epee::string_tools::buff_to_hex_nodelimer(raw);
+
+    // Split the long string into multiple lines, similarly to how openssl
+    // prints certificates.
+    for (size_t i = 63; i < h.length(); i += 64)
+    {
+      h.insert(i, 1, '\n');
+    }
+
+    h.insert(h.end(), '\n');
+    
+    return epee::file_io_utils::save_string_to_file(path_to_file, h);
+  }
+  else
+  {
+    return epee::file_io_utils::save_string_to_file(path_to_file, raw);
+  }
+}
+//----------------------------------------------------------------------------------------------------
+
+bool wallet2::load_from_file(const std::string& path_to_file, std::string& target_str,
+                             size_t max_size) const
+{
+  std::string data;
+  bool r = epee::file_io_utils::load_file_to_string(path_to_file, data, max_size);
+  if (!r)
+  {
+    return false;
+  }
+
+  // Ignore any whitespace.
+  // We try to convert the version without whitespace from hex to binary.
+  // If that fails, we stick to the original data (hence the copy). 
+  std::string without_whitespace(data);
+  without_whitespace.erase(
+    std::remove_if(
+      without_whitespace.begin(), 
+      without_whitespace.end(), 
+      [](char c){ return (c =='\r' || c =='\t' || c == ' ' || c == '\n');}),
+    without_whitespace.end());
+
+  r = epee::string_tools::validate_hex(without_whitespace.size(), without_whitespace);
+  if (!r)
+  {
+    target_str = data;
+    return true;
+  }
+
+  std::string res;
+  r = epee::string_tools::parse_hexstr_to_binbuff(without_whitespace, res);
```

Comment:
What I had in mind was more trying to parse, and then falling back to checking for hex. The patch checks whether the input is valid hex first, which might have some false positives. Admittedly probably not many, but it depends on what is being saved. Like for instance one thing above is saving a hex dump of something, so loading one of those files saved with current monero would actually load the wrong data.
In any case, if we decide it's not worth bothering about, I wonder whether this should go to the epee code itself, as an optional parameter. It might be useful elsewhere.

Reply:
Problem with trying to parse the binary data first is that this function has no idea how to do that.  We'd need to either supply a callback (convoluted and difficult to understand) or update all code that reads from files, which is error-prone and easy to forget in the future.

The mentioned case is IMO a bug, I think we should not convert to hex what's already ASCII. Fixing it in that place.

As for moving the whole implementation to epee, honestly, I've thought it's a third-party library - pushing to upstream and maintaining a fork in the meantime is pretty annoying. If we can just edit epee, it might be a better option.

Reply:
It's only a problem if you make the decision inside the read function :)
epee is only used in monero AFAIK.

Reply:
This question has been worked around; marking this conversation as resolved.